File: Libraries/Json/native/lang-javascript/source/json.js

LIB$json$parseJson = function(globals, rawText) {
    try {
        return LIB$json$convertJsonThing(globals, window.JSON.parse(rawText));
    } catch (e) {
        return null;
    }
};

LIB$json$convertJsonThing = function(globals, thing) {
    var type = LIB$Json$typeClassifyHelper(thing);
    switch (type) {
        case 'null': return buildNull(globals);
        case 'bool': return buildBoolean(globals, thing);
        case 'string': return buildString(globals, thing);
        case 'list':
            var list = [];
            for (i = 0; i < thing.length; ++i) {
                list.push(LIB$json$convertJsonThing(globals, thing[i]));
            }
            return buildList(list);
        case 'dict':
            var keys = [];
            var values = [];
            for (var rawKey in thing) {
                keys.push(rawKey);
                values.push(LIB$json$convertJsonThing(globals, thing[rawKey]));
            }
            return buildStringDictionary(globals, keys, values);
        case 'int':
            return buildInteger(globals, thing);
        case 'float':
            return buildFloat(globals, thing);
        default:
            return buildNull(globals);
    }
};

LIB$Json$typeClassifyHelper = function(t) {
    if (t === null) return 'null';
    if (t === true || t === false) return 'bool';
    if (typeof t == "string") return 'string';
    if (typeof t == "number") {
        if (t % 1 == 0) return 'int';
        return 'float';
    }
    ts = Object.prototype.toString.call(t);
    if (ts == '[object Array]') {
        return 'list';
    }
    if (ts == '[object Object]') {
        return 'dict';
    }
    return 'null';
};

File: Libraries/Json/native/lang-python/source/json.py

def lib_json_parse_impl(g, raw):
  import json
  try:
    return lib_json_parse_json_thing(g, json.loads(raw))
  except:
    return None

def lib_json_parse_json_thing(g, item):
  if item == None: return buildNull(g)
  t = str(type(item))
  if "'bool'" in t:
    return buildBoolean(g, item)
  if "'int'" in t or "'long'" in t:
    return buildInteger(g, item)
  if "'float'" in t:
    return buildFloat(g, item)
  if "'str'" in t or "'unicode'" in t:
    return buildString(g, str(item))
  if "'list'" in t:
    output = []
    for o in item:
      output.append(lib_json_parse_json_thing(g, o))
    return buildList(output)
  if "'dict'" in t:
    keys = []
    values = []
    for key in item.keys():
      keys.append(key)
      values.append(lib_json_parse_json_thing(g, item[key]))
    return buildStringDictionary(g, keys, values)
  return buildNull(g);
